[RFC] ath79: add encrypted Alpha Networks factory image for D-Link devices

raenye opened the pull request:
Currently, OpenWrt generates factory images for several D-Link
devices manufactured by Alpha Networks (in particular: DIR-842,
DIR-859, DAP-1720, DIR-X1860) that must be flashed via the D-Link web
recovery mode. This is cumbersome, as devices in recovery mode offer no
DHCP, and not all web browsers upload the firmware as required by
the recovery webserver ('application/octet-stream' binary form-data).

To use the vendor upgrade firmware web UI for installation, the firmware
needs to be encrypted using AES-256-CBC (with non-standard paddings and
no salt). The AES key and IV are created by mangling the key and IV
hard-coded in the GPL sources officially released by D-Link (see, e.g.,
`DIR842C3_GPL312EUb01DLink/tools/encimg/fw_sign_data.c`), and can also be
extracted from the official firmware image via `strings /htdocs/cgibin`.

 The mangling process is described by the following code excerpt:
```
 python -c 'import sys; s,k,i=sys.argv[1:]; X=lambda x: bytes(
   (j+1)^ord(a)^ord(b) for j,(a,b) in enumerate(zip(x,9*s))
   ).hex(); print(X(k),X(i))' $(SEAMA_SIGNATURE) $(32B_KEY) $(16B_IV)
```
 In fact, this is the firmware encryption method implemented by
 firmware-utils' nec-enc. An equivalent is implemented in Makefile as `alpha-xor`.

 * Rename current `factory.bin` images to `recovery.bin`
 * Add new `factory.bin` image (AES-encrypted `recovery.bin`)

Installation:
 * Web UI: http://192.168.0.1, flash `factory.bin`
 * Recovery Web UI: Keep reset button pressed during power-on,
   flash `recovery.bin` via http://192.168.0.1

Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>
Signed-off-by: Rani Hod <rani.hod@gmail.com>

raenye commented:
@s-2 please have a look.

I'm not sure which approach is better (both 1 and 2 are currently implemented):
1. To use `alpha_imgenc`. This requires adding `alpha_encimg.c` to `firmware-utils` (here this is done via `tools/firmware-utils/patches/010-add-alpha_imgenc.patch`); or
2. To manually pad and then invoke `openssl` cli in `generic.mk` with manually mangled key/iv; or
3. To manually pad and then invoke `openssl` cli in `generic.mk` with key/iv mangled by `nec-enc`. This requires modifying `nec-enc.c` to (i) allow stdin/stdout for input/output, and (ii) increase `KEY_SIZE` limit from 16 to 32(?).

This is mostly untested, but I checked that for DAP-1720 both implementations give the same output (`factory.bin` equals `factory2.bin`). I'd appreciate it if you could verify it works for at least one of your devices.

Another question: for DIR-842-C3 -- the one that needs `$(SEAMA_SIGNATURE)EU` -- why can't you just redefine `$(SEAMA_SIGNATURE)` to include the `EU`? (i.e. does the web recovery even check that the signature has no `EU`)

s-2 commented:
> Another question: for DIR-842-C3 -- the one that needs `$(SEAMA_SIGNATURE)EU` -- why can't you just redefine `$(SEAMA_SIGNATURE)` to include the `EU`? (i.e. does the web recovery even check that the signature has no `EU`)

For the C1 and C2 variant, the encryption was added only after the device was already on the market (which is why D-Link offer a separate `MiddleImage` firmware to upgrade older devices in two steps, i.e. the older image is the same format as the bootloader);
I can't recall if C3 EU was still published with unencrypted firmware available as well (will check the downloads for the existence of MiddleFirmware), but if someone still uses an older device, the (unencrypted recovery) image using EU signature might not be accepted in the regular OEM updater. I have not checked this yet, but if that happens the user could still flash D-Link MiddleFirmware to the latest OEM, and then upgrade using OpenWrt factory (with -EU signature) from there, so this should be fine (just needs to be well-documented).

Regarding the third option: I think if we would have to touch `nec-enc.c`, the only decent way to do so would be merging the functionality with the new tool (I had even thought about that before; actually, the `encimg` binary found in the D-Link filesystem has a parameter that allows using the old-style xor for the full image).

After all, we should probably go for the 2nd option, possibly without the separate signatures for the C3 EU variant, I will dig out my C3 (hopefully over the weekend) and do some more testing :slightly_smiling_face: 
In the meantime, since you already have a DAP-1720 at hand right now, could you try what happens if you specify a different signature for `seama-seal` (just append `foo` or something) and flash via the recovery?

raenye commented:
> could you try what happens if you specify a different signature for `seama-seal`

I'll try and report back.

> DIR-X1860

BTW, I couldn't find it in the codebase (ath79, ramips, ...)

raenye commented:
There's also something wrong about the commits. The fact that I used two of your commits (one of which was heavily modified) makes github say stuff like "_s-2 added 2 commits 13 hours ago_".

s-2 commented:
> BTW, I couldn't find it in the codebase (ath79, ramips, ...)

it's ramips mt7621, but not in OpenWrt yet. There has been a commit in `nbd`´s staging tree for many years, and [a forum thread](https://forum.openwrt.org/t/106500) I had started a few years ago (where someone eventually mentioned the existence of that commit by nbd), but I had not spent further time on it back then (initially could not get the kernel running, since relocation was needed etc., I was not familair with FIT-based images back then), also considering there was no progress on the encryption PR for ~1.5 years.



> The fact that I used two of your commits

it's probably just the sign-off, you can add yours as well or just mention me as co-authored-by etc., not sure if the order also makes a difference here.

s-2 commented:
Okay I've checked on DIR-842 C3 EU, the bootloader accepts both the `EU` and the non-EU signature (even though the resulting image offset is moved by 4 bytes when signature is 2 bytes longer, it's padded to 4 bytes).
Actually, looking at current D-Link firmwares for that device, the `MiddleFirmware` even has this appended, so it was probably quite pointless for me to verify this once again in the first place... :innocent: 
I haven't tested reverting the device to very old OEM firmware that require an unencrypted update file, but considering the D-Link MiddleFirmware has EU appended, I would expect this to be no issue as well, since using that image is the intended upgrade path anyways...

Thanks for spotting this, I guess it makes supporting this device even less complex :slightly_smiling_face:

csharper2005 commented:
@raenye why didn't you add your own SoB to the first commit?

raenye commented:
I just cherry picked this commit from Sebastian's tree as-is. But it appears it won't be needed, so I'll drop it altogether once I get to rework the PR draft.

I'd love to hear your opinion about the three options, BTW.

raenye commented:
@s-2 I pushed a version without the `seama-seal` commit, but I'm not sure I did good with the makefile variables. Could you please have a look? thanks.

raenye commented:
@hauke @RolandoMagico would love your opinion too about the best alternative of the three. Thanks.

DragonBluep commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1242,6 +1287,15 @@ define Device/dlink_dir-859-ax
+  IMAGES += factory.bin factory2.bin recovery.bin
+  IMAGE/recovery.bin := $$(IMAGE/default) | pad-rootfs -x 64 | seama | \
+	seama-seal | check-size
+  IMAGE/factory.bin := $$(IMAGE/recovery.bin) | \
```
Even the latest firmware of dir-859 is still the legacy seama header. Are you sure we need this?
https://media.dlink.eu/support/products/dir/dir-859/driver_software/dir-859_fw_reva_1-05b03_eu_multi_20161102.zip

raenye replied:
I'm not sure what is the bottom line. What should be changed there?

s-2 replied:
Nothing to change here, I just confirmed the encrypted image is indeed required for this device, since v1.07 is (among) the latest OEM firmwares available, and only accepts encrypted images (while 1.05 was still using the unencrypted (recovery) format).

raenye commented:
> Looks good, I can build and test for a real device, but you seem to generate two factory images anyways and they are identical already with both implementations?

I'm just not sure second modification of $(SEAMA_SIGNATURE) works correctly. But if you get identical `factory.bin` and `factory2.bin` for C3 too, it means the correct (EU) signature is used indeed.

> pad 16 bytes even if already aligned

I copied this behavior from `encimg.c`, and this is actually the only behavior that matters since `rootfs` is padded to 64KiB. Are you suggesting to avoid this padding altogether?

> It had been verified with so many different devices that I did not want to touch the padding logic anymore at that point, so just disabled it in openssl and kept doing it manually. But maybe for these devices we could even have openssl do it for us, will have to check the docs later when I'm back home.

By default, `openssl cli` uses PKCS 7 (IIRC) padding, which means adding 1 ≤ k ≤ 16 bytes, all of value k. In particular, for already aligned plaintexts it adds a line of 16 `0x10`, while we need 16 zeros. So I use `-nopad` and add 16 zeros via `dd`.
Maybe it doesn't even matter -- perhaps padding is discarded without being checked.

> By the way, appreciate your Python code golf for mangling the keys 🙂 

Thanks :smile:

> Reading that old forum thread on DIR-842 again, I think we should later add some simple decryption instructions to the wiki as well, so that regular users can also restore their devices to OEM firmware if needed, without the need for special tools.

I recall I once saw a Python version of `encimg`, maybe even by you. Do you know it? did it support decrypting?

s-2 commented:
> In particular, for already aligned plaintexts it adds a line of 16 `0x10`

Thanks, wasn't aware of that, so we will just stick with manual padding here :slightly_smiling_face:
When initially reversing `encimg` with ghidra, this is also they behaviour I found they had implemented manually, so I went to just do it the same way back then.


> I recall I once saw a Python version of `encimg`, maybe even by you

Indeed the first implementation in PR #3109 was in Python (you can still view it by clicking on individual commit hashes form the forced-push notifications, e.g. https://github.com/openwrt/openwrt/commit/45d0184b715f997738c965a262c36c48213ea653)
I had to rewrite it in C since there was no `pycrypto` installed on the buildbots...

I think I had only included the encryption, and checking my local disk I apparently never really implemented it in Python, but used the tool extracted from OEM firmware for decryption (tediously matched with the correct `.so` modules also from the firmware)

Regarding the wiki, we can probably just add the final `openssl` calls included the mangled keys for each device, which does not require the user to install pycrptodome or anything :slightly_smiling_face:

Will hopefully find time to build and test on DIR-842 EU tomorrow, but it should work when images are identical (and the bootloader also accepted both signatures, with and without `EU` appended).

Some time ago, I had even considered to make a generic Python framework (and finally GUI tool for the average users) for re-packing firmware in various formats... think of something like a spreadsheet for binary files, with abstract header fields defined (e.g. checksums, hashes, encrypted parts) based on other image components etc., so there would be just an image format model declaration for each device, and an analyzer / composer would use that to build / unpack images :slightly_smiling_face: (but did not really invest much time eventually, too many design choices to be made, just ending up procrastinating :innocent: )

s-2 commented:
Tested both `recovery.bin` and `factory.bin` flashing on DIR-842 C3 EU (shipped with bootloader as of v`3.10`), works fine, both factory images were also identical in my build :slightly_smiling_face:

s-2 commented:
@raenye any update on this, can we set this from Draft state to Open?
I just sifted through my piles of devices and found DIR-X1860 does not have a PR yet, and it could use this to build factory images 🙂

raenye commented:
I'm fine with both implementations, and would love some committers to express their opinion about which is better. TIA.
@hauke @ynezz @Ansuel @PolynomialDivision @Noltari

raenye committed:
ath79: add encrypted Alpha Networks factory image for D-Link devices

Currently, OpenWrt generates factory images for several D-Link
devices manufactured by Alpha Networks (in particular: DIR-842,
DIR-859, DAP-1720, DIR-X1860) that must be flashed via the D-Link web
recovery mode. This is cumbersome, as devices in recovery mode offer
no DHCP, and not all web browsers upload the firmware as required by
the recovery webserver (`application/octet-stream` binary form-data).

To use the vendor upgrade firmware web UI for installation, the firmware
needs to be encrypted using AES-256-CBC (with non-standard padding and
no salt). The AES key and IV are created by mangling the key and IV
hard-coded in the GPL sources officially released by D-Link (see, e.g.,
DIR842C3_GPL312EUb01DLink/tools/encimg/fw_sign_data.c), and can also be
extracted from the official firmware image via `strings /htdocs/cgibin`.
The mangling algorithm, implemented by `alpha-xor`, is the same as the
firmware encryption method `nec-enc` from firmware-utils.

* Rename current `factory.bin` images to `recovery.bin`
* Add new 'factory.bin' (= AES-encrypted `recovery.bin`)

Installation:

 * Web UI: http://192.168.0.1/UpdateFirmware.html, flash `factory.bin`
 * Recovery Web UI: Keep reset button pressed during power-on until red
   led is bright, release reset (the led will then blink slowly),
   set static IP to 192.168.0.10, open http://192.168.0.1 and upload
   `recovery.bin`

Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>
Signed-off-by: Rani Hod <rani.hod@gmail.com>

s-2 commented:
Awesome, thanks!
Considering this would indeed be used by multiple targets (was not even aware DIR-859 was moved to `tiny`, probably the same will happen for DIR-842 soon :thinking: ), maybe it could go into `include/image-commands.mk`, there are already several recipes for images over there.
I'll dig my DIR-X1860 A1 back out and update my branch, lots of dts changes since back then 😅

raenye commented:
OF seems to be an OpenWrt fork, maybe the contents of the following file could help.
(extracted rootfs squashfs via `binwalk -e`)

It seems to be looking for a particular JSON field (`alpha.fw_sign`) in the metadata.

```
$ cat squashfs-root/etc/scripts/fw_check.sh
#Check signature and checksum of upload firmware
#If checksum of upload firmware is wrong, the action "fwtool -q -i ..." would fail.
. /usr/share/libubox/jshn.sh
if ! fwtool -q -i /tmp/sysupgrade.meta "$1"; then
        echo -1
else
        json_load "$(cat /tmp/sysupgrade.meta)"
        json_select alpha
        json_get_var upload_fw_sign fw_sign
        echo "$upload_fw_sign" > /var/run/upload_fw_sign

        curr_fw_sign="`cat /rom/etc/alpha/fw_sign`" # wrgax10_dlink_dirx1860
        if [ "$upload_fw_sign" == "$curr_fw_sign" ]; then
                echo 0
        else
                echo -2
        fi
```
